### PR TITLE
Catalogs blogpost

### PR DESCRIPTION
**Pull request by dawidwys**

### CODE REVIEW

File: _posts/2020-07-21-catalogs.md [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+title: "Sharing is caring - Catalogs in Flink SQL"
+date: 2020-07-21T08:00:00.000Z
+categories: news
+authors:
+- dawid:
+  name: "Dawid Wysakowicz"
+  twitter: "dwysakowicz"
+---
+
+It's not a surprise that, in an era of digitalization, data is the most valuable asset in many companies: it's always the base for and product of any analysis or business logic. With an ever growing number of people working with data, it's a common practice for companies to build self-service platforms with the goal of democratising its access across different teams and — especially — to enable users from any background to be independent in their data needs. In such environments, metadata management becomes a crucial aspect. Without it, users often work blindly, spending too much time searching for datasets and their location, figuring out data formats and similar cumbersome tasks.
```
Comment (MarkSfik) [inline _posts/2020-07-21-catalogs.md]:
```suggestion
It's not a surprise that, in an era of digitalization, data is the most valuable asset in many companies: it's always the base for — and product of — any analysis or business logic. With an ever-growing number of people working with data, it's a common practice for companies to build self-service platforms with the goal of democratizing their access across different teams and — especially — to enable users from any background to be independent in their data needs. In such environments, metadata management becomes a crucial aspect. Without it, users often work blindly, spending too much time searching for datasets and their location, figuring out data formats and similar cumbersome tasks.
```

File: _posts/2020-07-21-catalogs.md [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+
+It is a common practice for companies to start building a data platform with a metastore, catalog, or schema registries of some sort in place. Those let you clearly separate making the data available from consuming it. That separation has a few benefits:
```
Comment (MarkSfik) [inline _posts/2020-07-21-catalogs.md]:
```suggestion
Frequently, companies start building a data platform with a metastore, catalog, or schema registry of some sort already in place. Those let you clearly separate making the data available from consuming it. That separation has a few benefits:
```

File: _posts/2020-07-21-catalogs.md [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+* improved productivity - The most obvious one. Making data reusable and shifting the focus on building new models/pipelines rather than data cleansing and discovery.
+* security - You can control the access to certain features of the data. For example, you can make the schema of dataset publicly available, but limit the actual access to the underlying data to only particular teams.
```
Comment (MarkSfik) [inline _posts/2020-07-21-catalogs.md]:
```suggestion
* security - You can control the access to certain features of the data. For example, you can make the schema of the dataset publicly available, but limit the actual access to the underlying data only to particular teams.
```

File: _posts/2020-07-21-catalogs.md [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+* compliance - If you have all the metadata in a central entity, it's much easier to ensure compliance with GDPR and other similar laws.
```
Comment (MarkSfik) [inline _posts/2020-07-21-catalogs.md]:
```suggestion
* compliance - If you have all the metadata in a central entity, it's much easier to ensure compliance with GDPR and similar regulations and legal requirements.
```

File: _posts/2020-07-21-catalogs.md [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+
+## What is stored in a catalog?
+
+Almost all data sets can be described by certain properties that must be known in order to consume them. Those include:
+* schema - It describes the actual contents of the data, what columns it has, what are the constraints (e.g. keys) on which the updates should be performed, which fields can act as time attributes, what are the rules for watermark generation and so on.
+* location - Does the data come from Kafka or a file in a filesystem? How do you connect to the external system? Which topic or file name do you use?
+* format - Is the data serialized as json, csv, or maybe avro records?
```
Comment (MarkSfik) [inline _posts/2020-07-21-catalogs.md]:
```suggestion
* format - Is the data serialized as JSON, CSV, or maybe Avro records?
```

File: _posts/2020-07-21-catalogs.md [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+* statistics - We can also store additional information that can be useful when creating an execution plan of our query. For example, we can choose the best join algorithm, based on the number of rows in joined datasets.
```
Comment (MarkSfik) [inline _posts/2020-07-21-catalogs.md]:
```suggestion
* statistics - You can also store additional information that can be useful when creating an execution plan of your query. For example, you can choose the best join algorithm, based on the number of rows in joined datasets.
```

Comment (MarkSfik) [reply _posts/2020-07-21-catalogs.md]: You used "you" to address the audience before, so I thought it is probably best to keep it consistent here. Just a thought :)

File: _posts/2020-07-21-catalogs.md [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+
+Catalogs don’t have to be limited to the metadata of datasets. You can usually store other objects that can be reused in different scenarios, such as:
+* functions - It's very common to have domain specific functions that can be helpful in different use cases. Instead of having to create them in each place separately, you can just create them once and share them with others.
+* queries - Those can be useful when you don’t want to persist a data set, but want to provide a recipe for creating it from other sources instead.
+
+## Catalogs support in Flink SQL
+Starting from version 1.9, Flink has a set of Catalog APIs that allow to integrate it with various catalog implementations. With the help of those APIs, you can query tables in Flink that were created in your external catalogs (e.g. Hive Metastore). Additionally, depending on the catalog implementation, you can create new objects such as tables or views from Flink, reuse them across different jobs, and possibly even use them in other tools compatible with that catalog. As of Flink 1.11, there are two catalog implementations supported by the community:
```
Comment (MarkSfik) [inline _posts/2020-07-21-catalogs.md]:
```suggestion
Starting from version 1.9, Flink has a set of Catalog APIs that allows to integrate Flink with various catalog implementations. With the help of those APIs, you can query tables in Flink that were created in your external catalogs (e.g. Hive Metastore). Additionally, depending on the catalog implementation, you can create new objects such as tables or views from Flink, reuse them across different jobs, and possibly even use them in other tools compatible with that catalog. As of Flink 1.11, there are two catalog implementations supported by the community:
```

File: _posts/2020-07-21-catalogs.md [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+* a comprehensive Hive catalog
+* a Postgres catalog (preview, read-only, for now)
+
+**Important:** Flink does not store data at rest; it is a compute engine and requires other systems to consume input from and write its output. This means that Flink does not own the lifecycle of the data. Integration with Catalogs does not change that. Flink uses catalogs for metadata management only.
+
+All you need to do to start querying your tables defined in either of these metastores is to create corresponding catalogs with connection parameters. Once this is done, you can use them the way you would in any relational database management system.
```
Comment (MarkSfik) [inline _posts/2020-07-21-catalogs.md]:
```suggestion
All you need to do to start querying your tables defined in either of these metastores is to create the corresponding catalogs with connection parameters. Once this is done, you can use them the way you would in any relational database management system.
```

File: _posts/2020-07-21-catalogs.md [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+--- create a catalog which gives access to the backing Hive installation
+CREATE CATALOG hive WITH (
+    'type'='hive',
+    'property-version'='1',
+    'hive-version'='2.3.6',
+    'hive-conf-dir'='/opt/hive-conf'
+);
+```
+
+After creating the catalogs, you can confirm that they're available to Flink and also list the databases or tables in each of these catalogs:
```
Comment (MarkSfik) [inline _posts/2020-07-21-catalogs.md]:
```suggestion
After creating the catalogs, you can confirm that they are available to Flink and also list the databases or tables in each of these catalogs:
```

File: _posts/2020-07-21-catalogs.md [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+  'topic'='lineitem',
+  'scan.startup.mode'='earliest-offset',
+  'properties.bootstrap.servers'='kafka:9092',
+  'properties.group.id'='testGroup',
+  'format'='csv',
+  'csv.field-delimiter'='|'
+);
+```
+
+With ``prod_lineitem`` stored in Hive, you can now write a query that will enrich the incoming stream with static data kept in Postgres. To illustrate how this works, let's calculate prices of items based on the current currency rates:
```
Comment (MarkSfik) [inline _posts/2020-07-21-catalogs.md]:
```suggestion
With ``prod_lineitem`` stored in Hive, you can now write a query that will enrich the incoming stream with static data kept in Postgres. To illustrate how this works, let's calculate the item prices based on the current currency rates:
```

File: _posts/2020-07-21-catalogs.md [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+JOIN prod_rates FOR SYSTEM_TIME AS OF l_proctime ON rs_symbol = l_currency
+WHERE
+  l_linestatus = 'O';
+```
+
+The query above uses a `SYSTEM AS OF` [clause]({{ site.DOCS_BASE_URL }}/dev/table/streaming/temporal_tables.html#temporal-table) for executing a temporal join. If you'd like to learn more about the different kind of joins you can do in Flink I highly encourage you to check [this documentation page]({{ site.DOCS_BASE_URL }}/dev/table/sql/queries.html#joins).
+
+## Conclusion
+
+Catalogs can be a powerful tool for building a platform where work of different teams can be made reusable. Centralizing the metadata is a common practice for improving the productivity, security, and compliance when working with data.
```
Comment (MarkSfik) [inline _posts/2020-07-21-catalogs.md]:
```suggestion
Catalogs can be extremely powerful when building data platforms aimed at reusing the work of different teams in an organization. Centralizing the metadata is a common practice for improving productivity, security, and compliance when working with data.
```

File: _posts/2020-07-21-catalogs.md [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+
+Flink provides a flexible metadata management capabilities, that aims at reducing the cumbersome, repetitive work needed before querying the data such as defining schemas, connection properties etc. As of version 1.11 Flink provides a native integration with Hive Metastore which is the most comprehensive one, and a read-only version for Postgres catalogs.
```
Comment (MarkSfik) [inline _posts/2020-07-21-catalogs.md]:
```suggestion
Flink provides flexible metadata management capabilities, that aim at reducing the cumbersome, repetitive work needed before querying the data such as defining schemas, connection properties etc. As of version 1.11, Flink provides a native, comprehensive integration with Hive Metastore and a read-only version for Postgres catalogs.
```

File: _posts/2020-07-21-catalogs.md [hunk 13]
```diff
@@ -0,0 +1,178 @@
...
+
+Get started with Flink and catalogs by reading [the docs]({{ site.DOCS_BASE_URL }}/dev/table/catalogs.html). And if you want to play around with Flink SQL (e.g. try out how catalogs work in Flink yourself), you can check [this demo](https://github.com/fhueske/flink-sql-demo) prepared by my colleagues Fabian and Timo — it runs in a dockerized environment, and I personally used it for the examples in this blog post.
```
Comment (MarkSfik) [inline _posts/2020-07-21-catalogs.md]:
```suggestion
You can get started with Flink and catalogs by reading [the docs]({{ site.DOCS_BASE_URL }}/dev/table/catalogs.html). If you want to play around with Flink SQL (e.g. try out how catalogs work in Flink yourself), you can check [this demo](https://github.com/fhueske/flink-sql-demo) prepared by my colleagues Fabian and Timo — it runs in a dockerized environment, and I personally used it for the examples in this blog post.
```

**Review by MarkSfik — COMMENTED** [review]
thanks for sharing this blog post @dawidwys !

I left some minor suggestions in the file. 

Please let me know if you have any questions! 

Thanks again! 
Markos

File: _posts/2020-07-21-catalogs.md [hunk 14]
```diff
@@ -0,0 +1,178 @@
+---
+layout: post
+title: "Sharing is caring - Catalogs in Flink SQL"
+date: 2020-07-21T08:00:00.000Z
+categories: news
```
Comment (morsapaes) [inline _posts/2020-07-21-catalogs.md]:
```suggestion
```

File: _posts/2020-07-21-catalogs.md [hunk 15]
```diff
@@ -0,0 +1,178 @@
...
+
+Frequently, companies start building a data platform with a metastore, catalog, or schema registry of some sort already in place. Those let you clearly separate making the data available from consuming it. That separation has a few benefits:
+* improved productivity - The most obvious one. Making data reusable and shifting the focus on building new models/pipelines rather than data cleansing and discovery.
+* security - You can control the access to certain features of the data. For example, you can make the schema of the dataset publicly available, but limit the actual access to the underlying data only to particular teams.
+* compliance - If you have all the metadata in a central entity, it's much easier to ensure compliance with GDPR and similar regulations and legal requirements.
```
Comment (morsapaes) [inline _posts/2020-07-21-catalogs.md]:
```suggestion
Frequently, companies start building a data platform with a metastore, catalog, or schema registry of some sort already in place. Those let you clearly separate making the data available from consuming it. That separation has a few benefits:

* **Improved productivity** - The most obvious one. Making data reusable and shifting the focus on building new models/pipelines rather than data cleansing and discovery.
* **Security** - You can control the access to certain features of the data. For example, you can make the schema of the dataset publicly available, but limit the actual access to the underlying data only to particular teams.
* **Compliance** - If you have all the metadata in a central entity, it's much easier to ensure compliance with GDPR and similar regulations and legal requirements.
```

File: _posts/2020-07-21-catalogs.md [hunk 16]
```diff
@@ -0,0 +1,178 @@
...
+
+## What is stored in a catalog?
+
+Almost all data sets can be described by certain properties that must be known in order to consume them. Those include:
+* schema - It describes the actual contents of the data, what columns it has, what are the constraints (e.g. keys) on which the updates should be performed, which fields can act as time attributes, what are the rules for watermark generation and so on.
+* location - Does the data come from Kafka or a file in a filesystem? How do you connect to the external system? Which topic or file name do you use?
+* format - Is the data serialized as JSON, CSV, or maybe Avro records?
+* statistics - You can also store additional information that can be useful when creating an execution plan of your query. For example, you can choose the best join algorithm, based on the number of rows in joined datasets.
```
Comment (morsapaes) [inline _posts/2020-07-21-catalogs.md]:
```suggestion
Almost all data sets can be described by certain properties that must be known in order to consume them. Those include:

* **Schema** - It describes the actual contents of the data, what columns it has, what are the constraints (e.g. keys) on which the updates should be performed, which fields can act as time attributes, what are the rules for watermark generation and so on.

* **Location** - Does the data come from Kafka or a file in a filesystem? How do you connect to the external system? Which topic or file name do you use?

* **Format** - Is the data serialized as JSON, CSV, or maybe Avro records?

* **Statistics** - You can also store additional information that can be useful when creating an execution plan of your query. For example, you can choose the best join algorithm, based on the number of rows in joined datasets.
```

File: _posts/2020-07-21-catalogs.md [hunk 17]
```diff
@@ -0,0 +1,178 @@
...
+
+Catalogs don’t have to be limited to the metadata of datasets. You can usually store other objects that can be reused in different scenarios, such as:
+* functions - It's very common to have domain specific functions that can be helpful in different use cases. Instead of having to create them in each place separately, you can just create them once and share them with others.
+* queries - Those can be useful when you don’t want to persist a data set, but want to provide a recipe for creating it from other sources instead.
```
Comment (morsapaes) [inline _posts/2020-07-21-catalogs.md]:
```suggestion
Catalogs don’t have to be limited to the metadata of datasets. You can usually store other objects that can be reused in different scenarios, such as:

* **Functions** - It's very common to have domain specific functions that can be helpful in different use cases. Instead of having to create them in each place separately, you can just create them once and share them with others.

* **Queries** - Those can be useful when you don’t want to persist a data set, but want to provide a recipe for creating it from other sources instead.
```

File: _posts/2020-07-21-catalogs.md [hunk 18]
```diff
@@ -0,0 +1,178 @@
...
+
+## Catalogs support in Flink SQL
+Starting from version 1.9, Flink has a set of Catalog APIs that allows to integrate Flink with various catalog implementations. With the help of those APIs, you can query tables in Flink that were created in your external catalogs (e.g. Hive Metastore). Additionally, depending on the catalog implementation, you can create new objects such as tables or views from Flink, reuse them across different jobs, and possibly even use them in other tools compatible with that catalog. As of Flink 1.11, there are two catalog implementations supported by the community:
+* a comprehensive Hive catalog
+* a Postgres catalog (preview, read-only, for now)
```
Comment (morsapaes) [inline _posts/2020-07-21-catalogs.md]:
```suggestion
Starting from version 1.9, Flink has a set of Catalog APIs that allows to integrate Flink with various catalog implementations. With the help of those APIs, you can query tables in Flink that were created in your external catalogs (e.g. Hive Metastore). Additionally, depending on the catalog implementation, you can create new objects such as tables or views from Flink, reuse them across different jobs, and possibly even use them in other tools compatible with that catalog. As of Flink 1.11, there are two catalog implementations supported by the community:

  1. A comprehensive Hive catalog

  2. A Postgres catalog (preview, read-only, for now)
```

File: _posts/2020-07-21-catalogs.md [hunk 19]
```diff
@@ -0,0 +1,178 @@
...
+
+**Important:** Flink does not store data at rest; it is a compute engine and requires other systems to consume input from and write its output. This means that Flink does not own the lifecycle of the data. Integration with Catalogs does not change that. Flink uses catalogs for metadata management only.
```
Comment (morsapaes) [inline _posts/2020-07-21-catalogs.md]:
```suggestion
<div class="alert alert-info" markdown="1">
<span class="label label-info" style="display: inline-block"><span class="glyphicon glyphicon-info-sign" aria-hidden="true"></span> Note</span>
Flink does not store data at rest; it is a compute engine and requires other systems to consume input from and write its output. This means that Flink does not own the lifecycle of the data. Integration with Catalogs does not change that. Flink uses catalogs for metadata management only.
</div>
```

Comment (morsapaes) [reply _posts/2020-07-21-catalogs.md]: Screen Shot 2020-07-21 at 09 12 07

File: _posts/2020-07-21-catalogs.md [hunk 20]
```diff
@@ -0,0 +1,178 @@
...
+);
+
+--- create a catalog which gives access to the backing Hive installation
+CREATE CATALOG hive WITH (
+    'type'='hive',
+    'property-version'='1',
+    'hive-version'='2.3.6',
+    'hive-conf-dir'='/opt/hive-conf'
+);
+```
```
Comment (morsapaes) [inline _posts/2020-07-21-catalogs.md]:
```suggestion
```sql
-- create a catalog which gives access to the backing Postgres installation
CREATE CATALOG postgres WITH (
    'type'='jdbc',
    'property-version'='1',
    'base-url'='jdbc:postgresql://postgres:5432/',
    'default-database'='postgres',
    'username'='postgres',
    'password'='example'
);

-- create a catalog which gives access to the backing Hive installation
CREATE CATALOG hive WITH (
    'type'='hive',
    'property-version'='1',
    'hive-version'='2.3.6',
    'hive-conf-dir'='/opt/hive-conf'
);
```
```

File: _posts/2020-07-21-catalogs.md [hunk 21]
```diff
@@ -0,0 +1,178 @@
...
+
+After creating the catalogs, you can confirm that they are available to Flink and also list the databases or tables in each of these catalogs:
+
+```
```
Comment (morsapaes) [inline _posts/2020-07-21-catalogs.md]:
```suggestion
```sql
```

File: _posts/2020-07-21-catalogs.md [hunk 22]
```diff
@@ -0,0 +1,178 @@
...
+ |-- c_phone: CHAR(15) NOT NULL
+ |-- c_acctbal: DOUBLE NOT NULL
+ |-- c_mktsegment: CHAR(10) NOT NULL
+ |-- c_comment: VARCHAR(117) NOT NULL
+```
+
+Now that you know which tables are available, you can write your first query.
+In this scenario, we keep customer orders in Hive (``dev_orders``) because of their volume, and reference customer data in Postgres (``prod_customer``) to be able to easily update it. Let’s write a query that shows customers and their orders by region and order priority for a specific day.
+
+```
```
Comment (morsapaes) [inline _posts/2020-07-21-catalogs.md]:
```suggestion
```sql
```

File: _posts/2020-07-21-catalogs.md [hunk 23]
```diff
@@ -0,0 +1,178 @@
...
+WHERE
+  FLOOR(o_ordertime TO DAY) = TIMESTAMP '2020-04-01 0:00:00.000'
+  AND NOT o_orderpriority = '4-NOT SPECIFIED'
+GROUP BY r_name, o_orderpriority
+ORDER BY r_name, o_orderpriority;
+```
+
+Flink's catalog support also covers storing Flink-specific objects in external catalogs that might not be fully usable by the corresponding external tools. The most notable use case for this is, for example, storing a table that describes a Kafka topic in a Hive catalog. Take the following DDL statement, that contains a watermark declaration as well as a set of connector properties that are not recognizable by Hive. You won't be able to query the table with Hive, but it will be persisted and can be reused by different Flink jobs.
+
+```
```
Comment (morsapaes) [inline _posts/2020-07-21-catalogs.md]:
```suggestion
```sql
```

File: _posts/2020-07-21-catalogs.md [hunk 24]
```diff
@@ -0,0 +1,178 @@
...
+  'properties.bootstrap.servers'='kafka:9092',
+  'properties.group.id'='testGroup',
+  'format'='csv',
+  'csv.field-delimiter'='|'
+);
+```
+
+With ``prod_lineitem`` stored in Hive, you can now write a query that will enrich the incoming stream with static data kept in Postgres. To illustrate how this works, let's calculate the item prices based on the current currency rates:
+
+```
```
Comment (morsapaes) [inline _posts/2020-07-21-catalogs.md]:
```suggestion
```sql
```

File: _posts/2020-07-21-catalogs.md [hunk 25]
```diff
@@ -0,0 +1,178 @@
...
+  l_currency AS `currency`,
+  rs_rate AS `cur_rate`,
+  (l_extendedprice * (1 - l_discount) * (1 + l_tax)) / rs_rate AS `open_in_euro`
+FROM hive.`default`.prod_lineitem
+JOIN prod_rates FOR SYSTEM_TIME AS OF l_proctime ON rs_symbol = l_currency
+WHERE
+  l_linestatus = 'O';
+```
+
+The query above uses a `SYSTEM AS OF` [clause]({{ site.DOCS_BASE_URL }}/dev/table/streaming/temporal_tables.html#temporal-table) for executing a temporal join. If you'd like to learn more about the different kind of joins you can do in Flink I highly encourage you to check [this documentation page]({{ site.DOCS_BASE_URL }}/dev/table/sql/queries.html#joins).
```
Comment (morsapaes) [inline _posts/2020-07-21-catalogs.md]:
```suggestion
The query above uses a `SYSTEM AS OF` [clause]({{ site.DOCS_BASE_URL }}flink-docs-release-1.11/dev/table/streaming/temporal_tables.html#temporal-table) for executing a temporal join. If you'd like to learn more about the different kind of joins you can do in Flink I highly encourage you to check [this documentation page]({{ site.DOCS_BASE_URL }}flink-docs-release-1.11/dev/table/sql/queries.html#joins).
```

File: _posts/2020-07-21-catalogs.md [hunk 26]
```diff
@@ -0,0 +1,178 @@
...
+
+## Conclusion
+
+Catalogs can be extremely powerful when building data platforms aimed at reusing the work of different teams in an organization. Centralizing the metadata is a common practice for improving productivity, security, and compliance when working with data.
+
+Flink provides flexible metadata management capabilities, that aim at reducing the cumbersome, repetitive work needed before querying the data such as defining schemas, connection properties etc. As of version 1.11, Flink provides a native, comprehensive integration with Hive Metastore and a read-only version for Postgres catalogs.
+
+You can get started with Flink and catalogs by reading [the docs]({{ site.DOCS_BASE_URL }}/dev/table/catalogs.html). If you want to play around with Flink SQL (e.g. try out how catalogs work in Flink yourself), you can check [this demo](https://github.com/fhueske/flink-sql-demo) prepared by my colleagues Fabian and Timo — it runs in a dockerized environment, and I personally used it for the examples in this blog post.
```
Comment (morsapaes) [inline _posts/2020-07-21-catalogs.md]:
```suggestion
You can get started with Flink and catalogs by reading [the docs]({{ site.DOCS_BASE_URL }}flink-docs-release-1.11/dev/table/catalogs.html). If you want to play around with Flink SQL (e.g. try out how catalogs work in Flink yourself), you can check [this demo](https://github.com/fhueske/flink-sql-demo) prepared by my colleagues Fabian and Timo — it runs in a dockerized environment, and I personally used it for the examples in this blog post.
```

**Review by morsapaes — COMMENTED** [review]
Thanks, @dawidwys ! Built the website to check how the blogpost looked and there were a couple of things out of place and/or that could be better highlighted. Also corrected some dead links.

File: _posts/2020-07-21-catalogs.md (hunk 14 above)
Comment (morsapaes) [reply _posts/2020-07-21-catalogs.md]: A lot of blogposts are erroneously under this category. This should only be used for actual announcements. 😉

File: _posts/2020-07-21-catalogs.md [hunk 27]
```diff
@@ -0,0 +1,188 @@
...
+* **Functions** - It's very common to have domain specific functions that can be helpful in different use cases. Instead of having to create them in each place separately, you can just create them once and share them with others.
+
+* **Queries** - Those can be useful when you don’t want to persist a data set, but want to provide a recipe for creating it from other sources instead.
+
+## Catalogs support in Flink SQL
+Starting from version 1.9, Flink has a set of Catalog APIs that allows to integrate Flink with various catalog implementations. With the help of those APIs, you can query tables in Flink that were created in your external catalogs (e.g. Hive Metastore). Additionally, depending on the catalog implementation, you can create new objects such as tables or views from Flink, reuse them across different jobs, and possibly even use them in other tools compatible with that catalog. As of Flink 1.11, there are two catalog implementations supported by the community:
+
+  1. A comprehensive Hive catalog
+
+  2. A Postgres catalog (preview, read-only, for now)
```
Comment (rmetzger) [inline _posts/2020-07-21-catalogs.md:52]: I don't have much prior knowledge about Flink SQL (maybe that makes me the ideal reader for this blog post draft):
Before this post, I though the catalog is just good for storing metadata about tables: I thought you can use the Postgres catalog for storing your Kafka+Avro or Filesystem+Parquet table definitions in Postgres Tables.
After reading this post, it seems that this is possible, but it is also possible to query data from Postgres (because the tables defined in postgres are available in Flink SQL)
Somehow I would expect the "Postgres catalog" to be called "Postgres connector" or "Postgres integration".

Maybe it makes sense to state more explicitly what you can and can not do with these catalogs?


Side question: Why is it Postgres-only, not generically JDBC?

File: _posts/2020-07-21-catalogs.md [hunk 28]
```diff
@@ -0,0 +1,188 @@
...
+layout: post
+title: "Sharing is caring - Catalogs in Flink SQL"
+date: 2020-07-21T08:00:00.000Z
+authors:
+- dawid:
+  name: "Dawid Wysakowicz"
+  twitter: "dwysakowicz"
+---
+
+It's not a surprise that, in an era of digitalization, data is the most valuable asset in many companies: it's always the base for — and product of — any analysis or business logic. With an ever-growing number of people working with data, it's a common practice for companies to build self-service platforms with the goal of democratizing their access across different teams and — especially — to enable users from any background to be independent in their data needs. In such environments, metadata management becomes a crucial aspect. Without it, users often work blindly, spending too much time searching for datasets and their location, figuring out data formats and similar cumbersome tasks.
```
Comment (rmetzger) [inline _posts/2020-07-21-catalogs.md]: Maybe it's just my personal taste, but "era of digitalization" and "data is the most valuable asset" sounds like some marketing wording, ringing some alarm bells. If this text would not be published on the Flink blog, I would stop reading here.

I guess the intention here is to motivate the relevance of the catalog integrations in Flink. Maybe a more technical motivation is better suited for the Flink blog? I find it quite exciting that you can use Flink as a universal query engine for data stored in Postgres or tables defined in the Hive Metastore.

**Review by rmetzger — APPROVED** [review]
I'm generally fine with publishing the blog post as-is, however, I left two high level comments for making it a bit stronger (in my opinion)

File: _posts/2020-07-21-catalogs.md (hunk 28 above)
Comment (dawidwys) [reply _posts/2020-07-21-catalogs.md]: I will try to rephrase it to be less "marketing".

File: _posts/2020-07-21-catalogs.md (hunk 27 above)
Comment (dawidwys) [reply _posts/2020-07-21-catalogs.md:52]: Background: I am not the author of the Postgres catalog and I did not directly participate in the design.

The idea of catalogs is that you could do both:
1. store Flink specific metadata
2. query the non-specific external data

Postgres Catalog implements only the latter.  In Flink a connector imo is rather well defined and means either source or sink that you can use for reading/writing data and thus I think it does not fit in here. Integration in my opinion is too broad and not well defined. The purpose of a catalog is to read/write/make use of metadata.

It is Postgres only because, it was implemented that way. I was also surprised when I was writing the blogpost and preparing the demo. I think there is a lot of potential for better unification the current implementation across different DBs.

In the post I tried to give a high level idea why you should think of catalogs when working with SQL and give an overview in a form of an e2e example what you can achieve in Flink. My intention was not to give a comprehensive overview of all available features. Nevertheless I am open for suggestion if you think a differently orientated post would make more sense.

File: _posts/2020-07-21-catalogs.md (hunk 28 above)
Comment (dawidwys) [reply _posts/2020-07-21-catalogs.md]: I though about it again this morning. I removed the first sentence with the "marketing" wording.

I did not further change the introduction. I understand your point of "using Flink as a universal query engine..." My idea for this blog was slightly different though, with more stress on the relevance of catalogs as such. That's the core idea I build around. Therefore the introduction is not Flink specific and less technical. Only once I explain the relevance of catalogs I go into the Flink catalogs support. This is to showcase that Flink does well on that important topic of catalogs that I explain in the beginning.

File: _posts/2020-07-21-catalogs.md (hunk 27 above)
Comment (dawidwys) [reply _posts/2020-07-21-catalogs.md:52]: I tried to elaborate more explicitly about the two approaches:
* integration
* metadata store

**Comment by dawidwys** [comment]
Let me know what you think about the latest changes @rmetzger @morsapaes @MarkSfik

File: _posts/2020-07-21-catalogs.md (hunk 27 above)
Comment (rmetzger) [reply _posts/2020-07-21-catalogs.md:52]: Thanks a lot for the explanation!

Maybe it would make sense to add a sentence to the beginning of the blog post that states the purpose of it, for example:
> In this blog post, we want to give a high level overview of catalogs in Flink, and give an overview in the form of an end to end example that you can try out yourself

This would help manage the expectation of the reader. I was somehow expecting a "deep dive" into the new Hive / Postgres catalogs

Comment (rmetzger) [reply _posts/2020-07-21-catalogs.md:52]: I should add that the blog post is generally good to go as is -- maybe it's just my personal taste, so please don't overthink my feedback.
I don't want to be the guy who comes in last minute and overthrows everything.

File: _posts/2020-07-21-catalogs.md (hunk 28 above)
Comment (rmetzger) [reply _posts/2020-07-21-catalogs.md]: See my other comment. I realized that my expectations about the blog post were not properly managed :)

File: _posts/2020-07-21-catalogs.md [hunk 29]
```diff
@@ -35,7 +35,12 @@ Catalogs don’t have to be limited to the metadata of datasets. You can usually
...
+
+  * Catalogs are sort of out-of-the box integration with an ecosystem such as RDBMs or Hive, where you can query the external, towards Flink, tables, views, or functions without additional connector configuration. The connector properties are automatically derived from the Catalog itself.
+  * A persistent store for Flink specific metadata. In this mode we additionally store connector properties alongside the logical metadata such as a schema or a name. That approach let's you store a full definition of e.g. a Kafka backed table with records serialized with Avro in Hive that can be later on used by Flink. However, as it incorporates Flink specific properties it can not be used by other tools that leverage Hive metastore.
```
Comment (morsapaes) [inline _posts/2020-07-21-catalogs.md]:
```suggestion
Starting from version 1.9, Flink has a set of Catalog APIs that allows to integrate Flink with various catalog implementations. With the help of those APIs, you can query tables in Flink that were created in your external catalogs (e.g. Hive Metastore). Additionally, depending on the catalog implementation, you can create new objects such as tables or views from Flink, reuse them across different jobs, and possibly even use them in other tools compatible with that catalog. In other words, you can see catalogs as having a two-fold purpose:

  * Provide an out-of-the box integration with ecosystems such as RDBMSs or Hive that allows you to query external objects like tables, views, or functions with no additional connector configuration. The connector properties are automatically derived from the catalog itself.
  
  * Act as a persistent store for Flink-specific metadata. In this mode, we additionally store connector properties alongside the logical metadata (e.g. schema, object name). That approach enables you to, for example, store a full definition of a Kafka-backed table with records serialized with Avro in Hive that can be later on used by Flink. However, as it incorporates Flink-specific properties, it can not be used by other tools that leverage Hive Metastore. 
```

**Comment by morsapaes** [comment]
I think this is clearer now, but would add a sentence to manage expectations in the opening paragraph, as Robert mentioned. Since I was aware of the context from the get-go, it was nice to have a more agnostic pair of eyes for review. 👀

File: _posts/2020-07-21-catalogs.md (hunk 27 above)
Comment (dawidwys) [reply _posts/2020-07-21-catalogs.md:52]: My personal take is that all feedback is valuable. Moreover I do realize I am not the best writer, but because I want to work on it, feedback can only help. ;)

File: _posts/2020-07-21-catalogs.md [hunk 30]
```diff
@@ -8,6 +8,10 @@ authors:
   twitter: "dwysakowicz"
 ---
 
+In this blog post, I want to give a high level overview of catalogs in Flink. I will describe why should you consider using one and what can you achieve with it in place. I will also try to showcase how easy it is to use catalogs in the form of an end to end example that you can try out yourself.
+
+## Why do I need a catalog?
+
 With an ever-growing number of people working with data, it's a common practice for companies to build self-service platforms with the goal of democratizing their access across different teams and — especially — to enable users from any background to be independent in their data needs. In such environments, metadata management becomes a crucial aspect. Without it, users often work blindly, spending too much time searching for datasets and their location, figuring out data formats and similar cumbersome tasks.
```
Comment (morsapaes) [inline _posts/2020-07-21-catalogs.md]: I'd probably place the intro sentence after this paragraph, then use the subtitle to kind of kick things off and transition to the actual blogpost. What do you think?

In general, it's nice to not speak in the first person in blogposts, so here's my proposal to slightly improve the sentence you wrote:

"In this blog post, we want to give you a high level overview of catalogs in Flink. We'll describe why you should consider using them and what you can achieve with one in place. To round it up, we'll also showcase how simple it is to combine catalogs and Flink, in the form of an end-to-end example that you can try out yourself."

**Comment by rmetzger** [comment]
+1 to merge.

**Comment by dawidwys** [comment]
Closed with dce4b86552b30e5d5e1490d131d27bbe7432ed72

**Comment by dawidwys** [comment]
Thank you all for the help with the post!
